sycl/test-e2e/VirtualFunctions/misc/group-barrier.cpp

```diff
@@ -0,0 +1,175 @@
+// REQUIRES: aspect-usm_shared_allocations
+//
+// On CPU it segfaults within the kernel that performs virtual function call.
+// https://github.com/intel/llvm/issues/15080
+// XFAIL: cpu
+// UNSUPPORTED: gpu
+// On GPU this test (its older version which used nd_item instead of group)
+// used to fail with UR_RESULT_ERROR_PROGRAM_LINK_FAILURE.
+// SPIR-V files produced by SYCL_DUMP_IMAGES could be linked just fine (using
+// both llvm-spirv -r + llvm-link and ocloc).
+// Current version hangs and therefore it is marked as unsupported to avoid
+// wasting time in CI and potentially blocking a machine.
+// Reported in https://github.com/intel/llvm/issues/15068
+//
+// This test checks that group operations (barrier in this case) work correctly
+// inside virtual functions.
+//
+// RUN: %{build} -o %t.out %helper-includes
+// RUN: %{run} %t.out
+
+#include <sycl/detail/core.hpp>
+#include <sycl/group_algorithm.hpp>
+#include <sycl/group_barrier.hpp>
+#include <sycl/usm.hpp>
+
+#include "helpers.hpp"
+
+#include <iostream>
+#include <numeric>
+
+namespace oneapi = sycl::ext::oneapi::experimental;
+
+class BaseOp {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable)
+  virtual int apply(int *, sycl::group<1>) = 0;
+};
+
+class SumOp : public BaseOp {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable)
+  virtual int apply(int *LocalData, sycl::group<1> WG) {
+    LocalData[WG.get_local_id()] += WG.get_local_id();
+    sycl::group_barrier(WG);
+    int Res = 0;
+    if (WG.leader()) {
+      for (size_t I = 0; I < WG.get_local_range().size(); ++I) {
+        Res += LocalData[I];
+      }
+    }
+
+    return sycl::group_broadcast(WG, Res);
```

Comment:
This likely contains another `group_barrier` inside. Would it make sense to change the code to store the leader's value in line 50, then have a `barrier` and then read leader's value in each of the WIs before returning?

Reply:
Right, `group_broadcast` implies `group_barrier`. Considering that the test is named `group-barrier`, I've replaced `group_broadcast` with "manual broadcast" in b948b36c3fc047343aaefd9defdb4a932b6c2fa6